pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,368 @@
+		ObjectMeta: metav1.ObjectMeta{Name: SVFeatureStateCRName, Namespace: namespace},
+		Spec: featurestatesv1alpha1.CnsCsiSvFeatureStatesSpec{
+			FeatureStates: featureStates,
+		},
+	}
+	err = controllerRuntimeClient.Create(ctx, cnsCsiSvFeatureStates)
+	if err != nil {
+		pendingCRUpdatesObj.pendingCRUpdate[namespace] = cnsCsiSvFeatureStates
+	}
+	return nil
+}
+
+// configMapUpdated is the call back function for config-map informer
+func configMapUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	newfssConfigMap, ok := oldObj.(*v1.ConfigMap)
+	if newfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized old object %+v", newObj)
+		return
+	}
+	oldfssConfigMap, ok := newObj.(*v1.ConfigMap)
+	if oldfssConfigMap == nil || !ok {
+		log.Warnf("configMapUpdated: unrecognized new object %+v", newObj)
+		return
+	}
+
+	if newfssConfigMap.Name == supervisorFeatureStatConfigMapName &&
+		newfssConfigMap.Namespace == supervisorFeatureStateConfigMapNamespace {
+		if reflect.DeepEqual(newfssConfigMap.Data, oldfssConfigMap.Data) {
+			return
+		}
+		var err error
+		var featureStates []featurestatesv1alpha1.FeatureState
+		for feature, state := range newfssConfigMap.Data {
+			var featureState featurestatesv1alpha1.FeatureState
+			featureState.Name = feature
+			featureState.Enabled, err = strconv.ParseBool(state)
+			if err != nil {
+				log.Errorf("failed to parse feature state value: %v for feature: %q", state, feature)
+				return
+			}
+			featureStates = append(featureStates, featureState)
+		}
+		for {
+			// if failed to create or update CR in the namespace, then error is not returned
+			// for such cases, CR update will be enqueued to pendingCRUpdatesObj
+			err := updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, featureStates)
+			if err != nil {
+				log.Errorf("failed to update feature states CR in the all workload namespaces. err: %v", err)
+			}
+		}
+	}
+}
+
+// configMapAdded adds feature state switch values from configmap that has been created on K8s cluster
+func namespaceAdded(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	namespace, ok := obj.(*v1.Namespace)
+	if namespace == nil || !ok {
+		log.Warnf("namespaceAdded: unrecognized object %+v", obj)
+		return
+	}
+	if _, ok = namespace.Labels[WorkLoadNamespaceLabelKey]; ok {
+		log.Infof("observed new workload namespace: %v", namespace.Name)
+		for {
+			// if failed to create CR in the namespace, then error is not returned
+			// for such case, CR update will be enqueued to pendingCRUpdatesObj
+			err := createFeatureStatesCRInWorkLoadNamespace(ctx, namespace.Name)
+			if err == nil {
+				break
+			}
+			log.Errorf("failed to create feature states CR in the workload namespace: %q, err: %v", namespace, err)
```

Comment:
Should we add a sleep here as you are retrying infinitely?

Reply:
done